### ISSUE TITLE
Potential restriction of fairlearn?

### ISSUE THREAD

**Issue by guanqun-yang** [Microsoft/fairlearn#4]
I am wondering if there are any underlying requisites to apply **any** sklearn model, when I applied `sklearn.neural_network.MLPClassifier()` it says
![fairlearn](https://user-images.githubusercontent.com/36497361/54322048-dca6f280-45b0-11e9-940e-c400c4017215.png)
But when I applied `sklearn.svm.SVC()`, this issues **doe not** arise. 
It seems to me that the input object should have `class_weight` attribute, which `MLPClassifier()` does not have. So is this true?

**Comment by MiroDudik** [comment]
That is indeed correct. See the documentation of expgrad:

    Required input arguments:
      ...
      learner -- a learner implementing methods fit(X,Y,W) and predict(X),
                 where X is the DataFrame of covariates, and Y and W
                 are the Series containing the labels and weights,
                 respectively; labels Y and predictions returned by
                 predict(X) are in {0,1}

But, it's a good idea to provide a more informative error message. Also,
maybe at some point in future, it might be nice to have wrapper that would
convert a standard (non-weighted) learner into a weighted learner.

**Comment by romanlutz** [comment]
I marked it as a feature-request. I cannot provide an estimate at this point, but will update as soon as this gets scheduled.

**Comment by romanlutz** [comment]
@Pooja1301 is looking into this now for `ExponentiatedGradient`, and then same resampling should be applicable to `GridSearch` subsequently. Stay tuned :-)

**Comment by Pooja1301** [comment]
I have started looking into it.

**Comment by romanlutz** [comment]
@violazhong @riedgar-ms and I discussed this earlier today. Capturing some of our notes.

The most important aspect of this work is passing in the resampling method. @riedgar-ms pointed out that this should probably be in the constructor or the `fit` method. Since we're passing the `estimator` in the constructor that seems to be the most suitable place IMO. Once we have that on the object we'd have to replace this line https://github.com/fairlearn/fairlearn/blob/d1581b4576fa5c4dbb0fb99bf5548c74ba05b0ac/fairlearn/reductions/_exponentiated_gradient/_lagrangian.py#L178
with something like this (pardon the pseudocode at times):
```
if estimator has fit function and fit function has sample_weight arg:
     estimator.fit(self.X, redY, **{self.sample_weight_name: redW})
elif user has provided resampling_method:
     resampled_X, resampled_Y = resampling_method(X, redY, redW)
     estimator.fit(resampled_X, resampled_Y)
else:
    raise exception and describe that we need sample_weight arg or resampling_method
```

I see a few different ways to do the resampling. I'm sure there are way more complicated ones as well.

### deterministic and keeping the same size
- scale weights so that the sum of all weights is equal to the number of rows in X
- add floor(scaled_weight) samples of each row to a new resampled dataset, e.g. if rescaled_weight=2.7 -> add 2
- order the residuals, so for our 2.7 that would be 0.7
- fill the remaining slots up starting from the largest residuals

Example:
```
dataset = [s1,s2,s3,s4,s5]
weights = [1, 2, 3, 4, 5]
scaled_weights = [1/3, 2/3, 1, 4/3, 5/3]
add one of samples s3, s4, and s5, then order residuals:
[1/3, 2/3, 0, 1/3, 2/3]
We have two slots left so those would go to samples s2 and s5 with the final resampled dataset:
[s3,s4,s5,s2,s5] (ordering doesn't matter)
```

### deterministic but not keeping the same size
- find smallest weight
- scale all weights up so that min = 1
- take `floor(weight)` and replicate each sample that many times

Example:
```
dataset = [s1,s2,s3,s4,s5]
weights = [0.1, 0.29, 1.3, 3.41, 9.54]
scaled_weights = [1, 2.9, 13, 34.1, 95.4]
final data set has s1 x 1, s2 x 2, s3 x 13, s4 x 34, s5 x 95
```
Of course, using `floor` is quite arbitrary here. This could just as well be cutting off at <0.5.

---

@Pooja1301  if you're still interested in picking up a part of this task, or a completely different one, please let me know and we can discuss. There are a ton of nice tasks coming up with plots that I believe will be fairly easy to get started on, so that should make it a good experience as a new contributor. Please let me know :-)

**Comment by MiroDudik** [comment]
I'm sorry--I'm a bit late to the discussion.

The deterministic strategy above would not be my first choice, because it's less sample efficient than the following simple randomized strategy guaranteed to always generate N data points:
* rescale the weights so that they sum to one
* put the examples left-to-right on a unit line segment; each example is represented by a line segment with the same length as its weight, so the first example begins at 0 and the last example ends at 1
* pick a random number z uniformly from the interval [0, 1/N]
* create a grid with points z, z+1/N, z+2/N, ... , z+(N-1)/N
* for each grid point select the example that it lends in

By sample efficient I mean: how large N do you need to guarantee that a resampling wrapper (applied to a sample-weights-unaware learning algorithm) returns a model that's epsilon-close to the model that optimizes the weighted objective.

What are the benefits of a deterministic strategy over randomized? One that I can think of is reproducibility, but that can be handled by a random seed.

Re. API for the wrapper, I was thinking about something like:
```python
class Resampler(BaseEstimator):
    def __init__(self, estimator, random_state=None, n=None): 
    ...
    
    def fit(self, X, y, sample_weights=None, **kwargs):
    ...
```
If `n=None`, it just means that the data set of the same size as provided will be generated. We could also factor the data set size by specifying a scaling factor relative to the provided data (and default to 1).

**Comment by romanlutz** [comment]
All of this sounds reasonable and is fairly close to @violazhong 's current implementation. @violazhong please let us know if any of this doesn't make sense. Happy to elaborate!

**Comment by adrinjalali** [comment]
Just a note that the solution would need to be a meta-estimator wrapping another estimator to pass the child estimator the modified samples. The scikit-learn API doesn't allow resampling as is, and the meta-estimator is the workaround.

**Comment by MiroDudik** [comment]
Re. meta-estimator... good point!

**Comment by violazhong** [comment]
@MiroDudik, could you provide more references (articles, papers, etc.) for the last two bullet points in your last comment? "epsilon-close" sounds like an important thing to make sure the resampling implementation does not mess up our reduction algorithm. 

@riedgar-ms  @romanlutz in case you are also interested...

**Comment by TamaraAtanasoska** [comment]
@adrinjalali does the metadata routing make this possible atm without additional implementation?

**Comment by adrinjalali** [comment]
In terms of implementation, if we take Miro's solution, sample weights are _consumed_ by this "meta-estimator" and not routed to the sub-estimator (since you don't want to double use sample weights in the process). Implementing this in `fit` doesn't require any special treatment, and I think it'd work out of the box.
